src/shell/shell_history.cpp

```diff
@@ -0,0 +1,163 @@
+/*
+ *  Copyright (C) 2023-2023  The DOSBox Staging Team
+ *
+ *  This program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License along
+ *  with this program; if not, write to the Free Software Foundation, Inc.,
+ *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+
+#include "shell.h"
+
+#include <fstream>
+
+#include "checks.h"
+#include "control.h"
+#include "dosbox.h"
+#include "fs_utils.h"
+#include "string_utils.h"
+
+CHECK_NARROWING();
+
+static constexpr int HistoryMaxLineSize = 256;
+static constexpr int HistoryMaxNumLines = 500;
+
+static std_fs::path get_shell_history_path();
+static bool command_is_exit(std::string_view command);
+
+ShellHistory& ShellHistory::Instance()
+{
+	static auto history = ShellHistory();
+	return history;
+}
+
+void ShellHistory::Append(std::string command, std::uint16_t code_page)
+{
+	auto to_utf8_str = [code_page](const std::string& dos_str) {
+		std::string utf8_str{};
+		dos_to_utf8(dos_str, utf8_str, code_page);
+		return utf8_str;
+	};
+	
+	trim(command);
+	
+	auto utf8_command = to_utf8_str(command);
+
+	if (!command_is_exit(command) && !command.empty() &&
+	    (commands.empty() || commands.back() != utf8_command)) {
+		commands.emplace_back(std::move(utf8_command));
+	}
+}
+
+std::vector<std::string> ShellHistory::GetCommands(std::uint16_t code_page) const
+{
+	auto to_dos_str = [code_page](const std::string& utf8_str) {
+		std::string dos_str{};
+		utf8_to_dos(utf8_str, dos_str, UnicodeFallback::Simple, code_page);
+		return dos_str;
+	};
+
+	std::vector<std::string> dos_encoded_commands{};
+	std::transform(commands.begin(),
+	               commands.end(),
+	               std::back_inserter(dos_encoded_commands),
+	               to_dos_str);
+
+	return dos_encoded_commands;
+}
+
+ShellHistory::ShellHistory() : path(get_shell_history_path())
+{
+	// Must check arguments directly as control->SwitchToSecureMode()
+	// will not be called until the first shell is run
+	if (control->arguments.securemode) {
+		return;
+	}
+	if (path.empty()) {
+		return;
+	}
+	auto history_file = std::ifstream(path);
```

Comment:
`const`? I won't make the same comment everywhere, but we try to maximise the use of `const`s where applicable 😎